### PR TITLE
feature/2198 settings page

### PR DESCRIPTION
**Pull request by jeppekroghitk**
#### Link to ticket

https://leantime.itkdev.dk/#/tickets/showTicket/2198

#### Description

- **Added settings page with input for ticket cache timeout**
- **Grabbing cache timeout value in controller and passing to template**
- **Settings hidden input with value from ticket cache timeout**
- **Grabbing value from hidden input, and setting cachetimeout const in javascript**

#### Screenshot of the result

Screenshot 2024-08-27 at 10 40 55

#### Checklist

- [ ] My code is covered by test cases.
- [ ] My code passes our test (all our tests).
- [ ] My code passes our static analysis suite.
- [ ] My code passes our continuous integration process.

If your code does not pass all the requirements on the checklist you have to add a comment explaining why this change 
should be exempt from the list.

#### Additional comments or questions

If you have any further comments or questions for the reviewer please add them here.

### CODE REVIEW

File: Controllers/Settings.php [hunk 1]
```diff
@@ -3,35 +3,59 @@
...
+
+    /**
+     * post method
+     * @param array<string, int> $params
+     * @return RedirectResponse
+     * @throws \Exception
+     */
+    public function post(array $params): RedirectResponse
+    {
+        $this->settingsRepo->saveSetting('timetablesettings.ticketscache', (int) ($params['ticketCacheExpiration'] ?? 0));
```
Comment (turegjorup) [inline Controllers/Settings.php:55]: Two things:
1. Given that we are writing to a shared repository, `timetablesettings` should probably be name spaced somehow to minimize any (even remote) risk of the setting being overwritten by either core or other plugins . E.g. `itk-leantime-timetable` 
2. We should use the same paramter name all round for the same value to minimize confusion. So if the param is `ticketCacheExpiration` then the settings key should be the same. So with (1) it should be `saveSetting('itk-leantime-timetable.ticketCacheExpiration', ...)`

File: Controllers/Settings.php [hunk 2]
```diff
@@ -3,35 +3,59 @@
...
+
+        $this->tpl->setNotification('The settings were successfully saved.', 'success');
```
Comment (turegjorup) [inline Controllers/Settings.php:57]: I like that failure is not an option 😜

File: Controllers/Settings.php [hunk 3]
```diff
@@ -3,35 +3,59 @@
...
-     * get
+     * get method
      *
      * @return Response
-     *
      * @throws \Exception
      */
     public function get(): Response
     {
+        $ticketCacheExpiration = (int) ($this->settingsRepo->getSetting('timetablesettings.ticketscache') ?: 1200);
```
Comment (turegjorup) [inline Controllers/Settings.php:40]: Se next comment about naming

File: Controllers/TimeTable.php [hunk 4]
```diff
@@ -108,6 +106,7 @@ public function get(): Response
         $userIdForFilter = null;
         $searchTermForFilter = null;
         $now = CarbonImmutable::now();
+        $ticketsCache = $this->settings->getSetting('timetablesettings.ticketscache') ?? 1200;
```
Comment (turegjorup) [inline Controllers/TimeTable.php:109]: Naming again: `ticketCacheExpiration`

File: Controllers/TimeTable.php [hunk 5]
```diff
@@ -167,6 +166,7 @@ public function get(): Response
         $this->tpl->assign('timesheetsByTicket', $timesheetsByTicket);
         $this->tpl->assign('weekDays', $days);
         $this->tpl->assign('weekDates', $weekDates);
+        $this->tpl->assign('ticketsCache', $ticketsCache);
```
Comment (turegjorup) [inline Controllers/TimeTable.php:169]: Naming again: `ticketCacheExpiration`

File: Templates/settings.blade.php [hunk 6]
```diff
@@ -2,14 +2,31 @@
...
 
 
-<div class="maincontent">
-    <div class="maincontentinner">
-        <h1>Hello World settings!</h1>
+    <div class="maincontent">
+        <?php if (isset($tpl)) {echo $tpl->displayNotification();}  ?>
+        <div class="maincontentinner">
+            <h5 class="subtitle">TimeTable Settings</h5>
+            <p style="padding-bottom: 15px">These settings will change the way the TimeTable plugin works.</p>
```
Comment (turegjorup) [inline Templates/settings.blade.php:14]: Not a "must fix" but.... 
Not a huge fan of inline styling or `px` values. Looks like Leantime is using bootstrap (?), so look at https://getbootstrap.com/docs/5.3/utilities/spacing/#margin-and-padding or add a class in the plugins css file

File: Templates/timetable.blade.php [hunk 7]
```diff
@@ -3,6 +3,7 @@
 @section('content')
     <div class="time-table-container">
     <input type="hidden" name="timetable-ticket-ids" value="{{$ticketIds}}" />
+    <input type="hidden" id="ticketCache" name="timetable-ticket-cache" value="{{$ticketsCache}}" />
```
Comment (turegjorup) [inline Templates/timetable.blade.php:6]: Naming again: `ticketCacheExpiration`

File: assets/timeTableApiHandler.js [hunk 8]
```diff
@@ -1,9 +1,10 @@
 /**
  * Class handles API requests for time table data.
  */
+const ticketCache = document.getElementById("ticketCache").value;
```
Comment (turegjorup) [inline assets/timeTableApiHandler.js:4]: Naming again: `ticketCacheExpiration` and namespacing given that any DOM id must be unique

File: assets/timeTableApiHandler.js [hunk 9]
```diff
@@ -1,9 +1,10 @@
...
 export default class TimeTableApiHandler {
   static cacheTimeouts = {
-    tickets: 1000,
+    tickets: parseFloat(ticketCache),
```
Comment (turegjorup) [inline assets/timeTableApiHandler.js:7]: Naming again: `ticketCacheExpiration`

**Review by turegjorup — APPROVED** [review]
Approved with comments. Nothing major - but namespacing and consistent naming would be nice.

**Comment by jeppekroghitk** [comment]
@turegjorup FYI im moving the implementation of your feedback to this PR, as i upgraded my local version of Leantime.

https://github.com/ITK-Leantime/leantime-timetable/pull/8

**Pull request merged**